{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gi|6273291|gb|AF191665.1|AF191665\n",
      "Seq('TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAA...AGA', SingleLetterAlphabet())\n",
      "902\n",
      "gi|6273290|gb|AF191664.1|AF191664\n",
      "Seq('TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAA...AGA', SingleLetterAlphabet())\n",
      "899\n",
      "gi|6273289|gb|AF191663.1|AF191663\n",
      "Seq('TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAA...AGA', SingleLetterAlphabet())\n",
      "899\n",
      "gi|6273287|gb|AF191661.1|AF191661\n",
      "Seq('TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAA...AGA', SingleLetterAlphabet())\n",
      "895\n",
      "gi|6273286|gb|AF191660.1|AF191660\n",
      "Seq('TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAA...AGA', SingleLetterAlphabet())\n",
      "893\n",
      "gi|6273285|gb|AF191659.1|AF191659\n",
      "Seq('TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAA...AGA', SingleLetterAlphabet())\n",
      "894\n",
      "gi|6273284|gb|AF191658.1|AF191658\n",
      "Seq('TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAA...AGA', SingleLetterAlphabet())\n",
      "896\n"
     ]
    }
   ],
   "source": [
    "########## Read in raw sequence in FASTA  ################\n",
    "\n",
    "from Bio import SeqIO\n",
    "for seq_record in SeqIO.parse(\"PKTest.fasta\", \"fasta\"):\n",
    "    print(seq_record.id)\n",
    "    print(repr(seq_record.seq))\n",
    "    print(len(seq_record))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The coding sequence is ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\n",
      "The template sequence is CTATCGGGCACCCTTTCAGCGGCCCATTACAATGGCCAT\n",
      "The mRNA sequence (of the coding sequence) is AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\n"
     ]
    }
   ],
   "source": [
    "######## Transcription #################\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "# 5' 3' DNA coding seqeunce (double strand)\n",
    "coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\", IUPAC.unambiguous_dna)\n",
    "print('The coding sequence is', coding_dna)\n",
    "\n",
    "#3' 5' DNA Template sequence (Double Strand)\n",
    "template_dna = coding_dna.reverse_complement()\n",
    "print('The template sequence is', template_dna)\n",
    "\n",
    "# T -> U RNA sequence (single strand)\n",
    "messenger_rna = coding_dna.transcribe()\n",
    "print('The mRNA sequence (of the coding sequence) is', messenger_rna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The first frame codons are ['ATG', 'GCC', 'ATT', 'GTA', 'ATG', 'GGC', 'CGC', 'TGA', 'AAG', 'GGT', 'GCC', 'CGA', 'TAG']\n",
      "The translation at each frame is ['MAIVMGR_KGAR_', 'WPL_WAAERVPD', 'GHCNGPLKGCPI']\n"
     ]
    }
   ],
   "source": [
    "##### DNA Translation ############   \n",
    "codonmatrix = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M', 'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K', 'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L', 'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q', 'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V', 'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E', 'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S', 'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_', 'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W'}\n",
    " \n",
    "# a function to translate a single codon\n",
    "def translate_codon(codon):\n",
    "    return codonmatrix.get(codon.upper(), 'x')\n",
    " \n",
    "# a function to split a sequence into codons\n",
    "def split_into_codons(dna, frame):\n",
    "    codons = []\n",
    "    for i in range(frame - 1, len(dna)-2, 3):\n",
    "        codon = dna[i:i+3]\n",
    "        codons.append(codon)\n",
    "    return codons\n",
    " \n",
    "# a function to translate a dna sequence in a single frame\n",
    "def translate_dna_single(dna, frame=1):\n",
    "    codons = split_into_codons(dna, frame)\n",
    "    amino_acids = ''\n",
    "    for codon in codons:\n",
    "        amino_acids = amino_acids + translate_codon(codon)\n",
    "    return amino_acids\n",
    " \n",
    "# a function to translate a dna sequence in 3 forward frames\n",
    "def translate_dna(dna):\n",
    "    all_translations = []\n",
    "    for frame in range(1,4):\n",
    "        all_translations.append(translate_dna_single(dna, frame))\n",
    "    return all_translations\n",
    "    \n",
    "split = split_into_codons('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG', 1)\n",
    "print('The first frame codons are', split)\n",
    "\n",
    "translation = translate_dna('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
    "print(\"The translation at each frame is\", translation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The first sequence similarity score is 6\n",
      "Score : 7\n",
      "3       10\n",
      " gatcgat\n",
      " gatcgat\n",
      "4       11\n",
      "n--------------------n\n",
      "Score : 7\n",
      "8           19\n",
      " attgatcgatc\n",
      " ttagatcgatc\n",
      "1           12\n",
      "n--------------------n\n",
      "Score : 7\n",
      "9         18\n",
      " ttgatcgat\n",
      " tagatcgat\n",
      "2         11\n",
      "n--------------------n\n",
      "Score : 8\n",
      "9          19\n",
      " ttgatcgatc\n",
      " tagatcgatc\n",
      "2          12\n",
      "n--------------------n\n",
      "Score : 7\n",
      "9           20\n",
      " ttgatcgatcg\n",
      " tagatcgatct\n",
      "2           13\n",
      "n--------------------n\n",
      "Score : 7\n",
      "9             22\n",
      " ttgatcgatcgat\n",
      " tagatcgatcttt\n",
      "2             15\n",
      "n--------------------n\n",
      "Score : 7\n",
      "10         19\n",
      " tgatcgatc\n",
      " agatcgatc\n",
      "3         12\n",
      "n--------------------n\n",
      "Score : 7\n",
      "11       18\n",
      " gatcgat\n",
      " gatcgat\n",
      "4       11\n",
      "n--------------------n\n",
      "Score : 8\n",
      "11        19\n",
      " gatcgatc\n",
      " gatcgatc\n",
      "4        12\n",
      "n--------------------n\n",
      "Score : 7\n",
      "11         20\n",
      " gatcgatcg\n",
      " gatcgatct\n",
      "4         13\n",
      "n--------------------n\n",
      "Score : 7\n",
      "11           22\n",
      " gatcgatcgat\n",
      " gatcgatcttt\n",
      "4           15\n",
      "n--------------------n\n",
      "Score : 7\n",
      "12           23\n",
      " atcgatcgatc\n",
      " ttagatcgatc\n",
      "1           12\n",
      "n--------------------n\n",
      "Score : 7\n",
      "12       19\n",
      " atcgatc\n",
      " atcgatc\n",
      "5       12\n",
      "n--------------------n\n",
      "Score : 7\n",
      "13         22\n",
      " tcgatcgat\n",
      " tagatcgat\n",
      "2         11\n",
      "n--------------------n\n",
      "Score : 8\n",
      "13          23\n",
      " tcgatcgatc\n",
      " tagatcgatc\n",
      "2          12\n",
      "n--------------------n\n",
      "Score : 7\n",
      "14         23\n",
      " cgatcgatc\n",
      " agatcgatc\n",
      "3         12\n",
      "n--------------------n\n",
      "Score : 7\n",
      "15       22\n",
      " gatcgat\n",
      " gatcgat\n",
      "4       11\n",
      "n--------------------n\n",
      "Score : 8\n",
      "15        23\n",
      " gatcgatc\n",
      " gatcgatc\n",
      "4        12\n",
      "n--------------------n\n",
      "Score : 7\n",
      "16       23\n",
      " atcgatc\n",
      " atcgatc\n",
      "5       12\n",
      "n--------------------n\n"
     ]
    }
   ],
   "source": [
    "############ Sequence alignment\n",
    "\n",
    "reference_sequence = 'actgatcgattgatcgatcgatcg'\n",
    "query_sequence   = 'tttagatcgatctttgatc'\n",
    "\n",
    "# here are the five bits of information we described before\n",
    "def score_match(subject, query, subject_start, query_start, length):\n",
    "    score = 0\n",
    "    # for each base in the match\n",
    "    for i in range(0,length):\n",
    "        # first figure out the matching base from both sequences\n",
    "        subject_base = subject[subject_start + i]\n",
    "        query_base = query[query_start + i]\n",
    "        # then adjust the score up or down depending on \n",
    "        # whether or not they are the same\n",
    "        if subject_base == query_base:\n",
    "            score = score + 1\n",
    "        else:\n",
    "            score = score - 1\n",
    "    return score\n",
    "    \n",
    "# here is the score for the match we were looking at above\n",
    "print(\"The first sequence similarity score is\", score_match(reference_sequence, query_sequence, 7, 4, 8))\n",
    " \n",
    "# the arguments are the five bits of information that define a match\n",
    "def pretty_print_match(subject, query, subject_start, query_start, length):\n",
    " \n",
    "    # first print the start/stop positions for the subject sequence\n",
    "    print(str(subject_start) + (' ' * length) + str(subject_start+length))\n",
    " \n",
    "    # then print the bit of the subject that matches\n",
    "    print(' ' + subject[subject_start:subject_start+length])\n",
    " \n",
    "    # then print the bit of the query that matches\n",
    "    print(' ' + query[query_start:query_start+length])\n",
    " \n",
    "    # finally print the start/stop positions for the query\n",
    "    print(str(query_start) + (' ' * length) + str(query_start+length))\n",
    " \n",
    "    print('n--------------------n')\n",
    "    \n",
    "    \n",
    "def try_all_matches(subject, query, score_limit):\n",
    "    for subject_start in range(0,len(subject)):\n",
    "        for query_start in range(0,len(query)):\n",
    "            for length in range(0,len(query)):\n",
    "                if (subject_start + length < len(subject) and query_start + length < len(query)):\n",
    "                    score = score_match(subject, query, subject_start, query_start, length)\n",
    "                    # only print a line of output if the score is better than some limie\n",
    "                    if (score >= score_limit):\n",
    "                        print(subject_start, query_start, length, score)\n",
    "                        \n",
    "                        \n",
    "def try_all_matches(subject, query, score_limit):\n",
    "    for subject_start in range(0,len(subject)):\n",
    "        for query_start in range(0,len(query)):\n",
    "            for length in range(0,len(query)):\n",
    "                if (subject_start + length < len(subject) and query_start + length < len(query)):\n",
    "                    score = score_match(subject, query, subject_start, query_start, length)\n",
    "                    # only print a line of output if the score is better than some limie\n",
    "                    if (score >= score_limit):\n",
    "                        print('Score : ' + str(score))\n",
    "                        pretty_print_match(subject, query, subject_start, query_start, length)\n",
    " \n",
    "try_all_matches(reference_sequence, query_sequence, 7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Paul\\AppData\\Local\\Continuum\\Anaconda3\\lib\\site-packages\\Bio\\pairwise2.py:993: BiopythonWarning: Import of C module failed. Falling back to pure Python implementation. This may be slooow...\n",
      "  'implementation. This may be slooow...', BiopythonWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The pairwise alignment for p53 is MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGP------\n",
      "MEESQSDISLELPLSQETFSGLWKLLPPEDIL-PSP-HCMDDLLL-PQDVEEFF---EGPSEALRV\n"
     ]
    }
   ],
   "source": [
    "#### Sequence alignment of p53\n",
    "from Bio import pairwise2\n",
    "from Bio.SubsMat import MatrixInfo as matlist\n",
    " \n",
    "matrix = matlist.blosum62\n",
    "gap_open = -10\n",
    "gap_extend = -0.5\n",
    " \n",
    "# Human and mouse amino acid sequence\n",
    "p53_h = \"MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGP\"\n",
    "p53_m = \"MEESQSDISLELPLSQETFSGLWKLLPPEDILPSPHCMDDLLLPQDVEEFFEGPSEALRV\"\n",
    " \n",
    "alns = pairwise2.align.globalds(p53_h, p53_m, matrix, gap_open, gap_extend)\n",
    " \n",
    "top_aln = alns[0]\n",
    "aln_human, aln_mouse, score, begin, end = top_aln\n",
    " \n",
    "print(\"The pairwise alignment for p53 is\", aln_human+'\\n'+aln_mouse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def ThanksCris(python):\n",
    "    return pythonclass.get(python.concepts(), 'Paul')"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
